### PR TITLE
ENG-9386: add resource AccessRules

### PR DESCRIPTION
**Pull request by ricardorey10**
## Description of the change

This PR adds a resource for the Gatekeeper's Access Rules.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Development

- [ ] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review

- [x] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [x] Jira issue referenced in commit message and/or PR title

### Testing
Did a lot of manual testing running `terraform apply`, `terraform import`, etc, and things look sane AFAICT. Acceptance tests are also passing.

### CODE REVIEW

File: cyral/resource_cyral_access_rules.go
Comment (Yowgf) [inline cyral/resource_cyral_access_rules.go:59]:
```suggestion
	return d.Set("rule", rules)
```

`d.Set` returns an error. Although we don't expect this to err, and we currently ignore this error in a lot of places in our code, I think it is best if we propagate it.

File: cyral/resource_cyral_access_rules.go
Comment (Yowgf) [inline cyral/resource_cyral_access_rules.go:140]: Maybe we can declare a variable like `repositoryID` before `d.SetId` and reuse it here and when creating the URL. Same for `user_account_id`.

File: cyral/resource_cyral_access_rules.go
Comment (Yowgf) [inline cyral/resource_cyral_access_rules.go:190]:
```suggestion
						panic("Failed to unmarshal access rules ID: " + err.Error())
```
I know we don't expect this to panic, but maybe it would be nice to include the error information, just in case.

File: cyral/resource_cyral_access_rules.go
Comment (Yowgf) [inline cyral/resource_cyral_access_rules.go:212]:
```suggestion
				Description: "ID of the repository.",
```
Just a minor detail (the dot).

FYI, the description is copied into the markdown that goes to the Provider docs, so you can write markdown syntax here if you wish :).

**Review by Yowgf — COMMENTED** [review]
The resource looks good, thanks Ricardo! I added some comments regarding some really minor details.

Also, when you finish the resource, you will want to generate the docs:

```
make docker-compose/docs
pre-commit run --show-diff-on-failure --color=always --all-files
```

In my machine, `make docker-compose/docs` is kind of bugged, it runs as a different user, so I need to run something like  `find docs -exec chown myuser {} \;` after it. If you could fix that it would be very much appreciated :).

File: cyral/resource_cyral_access_rules.go
Comment (ricardorey10) [reply cyral/resource_cyral_access_rules.go:59]: Done. Thanks for the suggestion!

File: cyral/resource_cyral_access_rules.go
Comment (ricardorey10) [reply cyral/resource_cyral_access_rules.go:140]: Done.

File: cyral/resource_cyral_access_rules.go
Comment (ricardorey10) [reply cyral/resource_cyral_access_rules.go:190]: Good idea. Done.

File: cyral/resource_cyral_access_rules.go
Comment (ricardorey10) [reply cyral/resource_cyral_access_rules.go:212]: Yeah, thanks for catching this, I was pretty inconsistent with punctuation. Done.

And thanks for the suggestion about markdown!

File: cyral/provider.go
Comment (VictorGFM) [inline cyral/provider.go:177]: Maybe its a good idea to have this defined near the group of the other `cyral_repository*` resources

**Comment by ricardorey10** [comment]
@Yowgf @VictorGFM @wcmjunior I still need to add unit tests, but I think the "core" logic should be good for review now.

Please let me know if I should also look into the "data_source" stuff, or if the resource is good enough for now.

File: cyral/resource_cyral_access_rules.go
Comment (VictorGFM) [inline cyral/resource_cyral_access_rules.go:140]: The resource ID is usually set after the request succeeds on creating the resource. This can be done in the WriteToSchema of the `AccessRulesResponse`.

File: cyral/resource_cyral_access_rules.go
Comment (VictorGFM) [inline cyral/resource_cyral_access_rules.go:191]: Instead of having to unmarshal the composed ID, maybe we could use the d.Get("repository_id").(string) and d.Get("user_account_id").(string) instead. I think this would simplify the code, but the way it is also works just fine.

File: cyral/resource_cyral_access_rules.go
Comment (VictorGFM) [inline cyral/resource_cyral_access_rules.go:209]: I think would be good to describe how the ID is composed (E.g.: {repository_id}/{user_account_id}), or we could just include a note on the documentation similar to what's done for the identity map resource (https://github.com/cyralinc/terraform-provider-cyral/blame/main/docs/resources/repository_identity_map.md#L5)). This would help users to understand how they should be importing the resource.

File: cyral/provider.go
Comment (VictorGFM) [reply cyral/provider.go:177]: We should also rename the file name to `resource_cyral_repository_access_rules.go` to match the resource name

File: cyral/resource_cyral_access_rules.go
Comment (VictorGFM) [inline cyral/resource_cyral_access_rules.go:277]: Would be good to set `MaxItems: 1` to the `config`.

**Review by VictorGFM — COMMENTED** [review]
At a high level looks very good @ricardorey10. I provided some initial comments that I think could help.

File: docs/resources/repository_access_rules.md
Comment (ricardorey10) [inline docs/resources/repository_access_rules.md:1]: This file still needs some work. In particular, I need to create a template file with instructions on how the id is calculated, like we do for https://github.com/cyralinc/terraform-provider-cyral/blob/main/docs/resources/repository_identity_map.md. My understanding is that this file is somehow built from a template. I'll have to look into this.

**Comment by ricardorey10** [comment]
@wcmjunior @VictorGFM @Yowgf Acceptance tests have not yet been written for this resource, because access rules live under user accounts, and we don't have the TF resource for user accounts yet.

Hannah was going to work on that, but she's out sick now, so I'll take that task in order to get this ready ASAP. Thus, I'll pause this PR for a while, finish the one for user accounts, and *then* get back to this one, once we have the ability to create user accounts via Terraform.

**Comment by wcmjunior** [comment]
Let's merge this changes only after we release `v2.9.0` (please reach out to me if you need more info) to avoid maintaining two production branches at this time. We will very likely have to keep a `v2` branch in the future, but I prefer to wait to avoid incoming cherry-picks. FYI @VictorGFM .

File: cyral/provider.go
Comment (ricardorey10) [reply cyral/provider.go:178]: Done both.

File: cyral/resource_cyral_access_rules.go
Comment (ricardorey10) [reply cyral/resource_cyral_access_rules.go:145]: The problem here is that the creation API does not return an ID, or any of the pieces we need to compute it. The only place where we have the full information is here (that's why Alex suggested I could set the ID here, IIRC).

File: cyral/resource_cyral_access_rules.go
Comment (ricardorey10) [reply cyral/resource_cyral_access_rules.go:196]: I feel like this is slightly cleaner, as we use the marshal function to compute the ID and then the unmarshal function to read it, so if it's OK then I'll just keep this as-is

File: cyral/resource_cyral_access_rules.go
Comment (ricardorey10) [reply cyral/resource_cyral_access_rules.go:214]: Yeah, I think having this in the docs precisely how we do it for that stuff you linked will be the cleanest way to do this.

File: cyral/resource_cyral_access_rules.go
Comment (ricardorey10) [reply cyral/resource_cyral_access_rules.go:282]: Did not think about that, but makes sense. Done.

**Comment by wcmjunior** [comment]
@ricardorey10 , please don 't forget to change the target to `main` instead.

File: cyral/resource_cyral_access_rules.go
Comment (ricardorey10) [reply cyral/resource_cyral_access_rules.go:145]: Ah, I keep forgetting what things we have to extract from the API call and which ones we get "for free". I'll fix this, thanks for the suggestion.

**Comment by ricardorey10** [comment]
@VictorGFM @Yowgf Quick poke about reviewing this :grin:

File: cyral/resource_cyral_repository_access_rules.go
Comment (VictorGFM) [inline cyral/resource_cyral_repository_access_rules.go:268]: We might want to make `valid_from` a required field depending on the conclusion of this [discussion](https://cyralinc.slack.com/archives/C0173L983EU/p1664240598935429).

**Review by VictorGFM — APPROVED** [review]
The changes look good Ricardo! I provided minor comments below that may require a few changes. Im already approving the PR though.

File: examples/resources/cyral_repository_access_rules/resource.tf
Comment (Yowgf) [inline examples/resources/cyral_repository_access_rules/resource.tf:1]: I am facing an infinite `terraform apply` loop when trying out this example, where even though my .tf file hasn't changed, the Terraform plan kept suggesting this change:

```
Terraform will perform the following actions:

  # cyral_repository_access_rules.t1 will be updated in-place
  ~ resource "cyral_repository_access_rules" "t1" {
        # (2 unchanged attributes hidden)

      ~ rule {
            # (2 unchanged attributes hidden)

          + config {}

            # (1 unchanged block hidden)
        }
      ~ rule {
          + config {}

            # (1 unchanged block hidden)
        }

        # (1 unchanged block hidden)
    }
```

Don't know how to fix it out of the top of my head though :/

**Review by Yowgf — COMMENTED** [review]
Looks good, Ricardo! Just some comments to improve the tests.

File: cyral/resource_cyral_repository_access_rules_test.go
Comment (ricardorey10) [reply cyral/resource_cyral_repository_access_rules_test.go:142]: For my own understanding: what is that used for, and why is it not needed for this test?

File: cyral/resource_cyral_repository_access_rules_test.go
Comment (ricardorey10) [reply cyral/resource_cyral_repository_access_rules_test.go]: Done.

File: cyral/resource_cyral_repository_access_rules_test.go
Comment (ricardorey10) [reply cyral/resource_cyral_repository_access_rules_test.go]: Sure thing, I can do this.

File: cyral/resource_cyral_repository_access_rules_test.go
Comment (ricardorey10) [reply cyral/resource_cyral_repository_access_rules_test.go]: Done.

File: examples/resources/cyral_repository_access_rules/resource.tf
Comment (ricardorey10) [reply examples/resources/cyral_repository_access_rules/resource.tf:1]: Can you share with me the contents of your `.tf` file?

File: cyral/resource_cyral_repository_access_rules_test.go
Comment (Yowgf) [reply cyral/resource_cyral_repository_access_rules_test.go:142]: When running an import test, the framework grabs the id of the resource referenced by the field ResourceName, and sends it over as a parameter to the importer.

ImportStatIdFunc is used to generate this id in a custom manner. It was used in the past in some resources where the actual ID format and import ID syntax were not the same, so we needed to grab for example the repository_id from the resource state and set that to be the id before passing it to the importer.

Comment (ricardorey10) [reply cyral/resource_cyral_repository_access_rules_test.go]: Gotcha. Thanks for the explanation. I thought I had deleted this code, as you suggested, but it looks like I actually forgot about it. Done with the deletion now.

File: examples/resources/cyral_repository_access_rules/resource.tf
Comment (Yowgf) [reply examples/resources/cyral_repository_access_rules/resource.tf:1]: I have pasted it below. It is essentially the configuration found in `examples/resources/cyral_repository_access_rules/resource.tf`.

```terraform
resource "cyral_repository" "t1" {
  name = "test1"
  type  = "sqlserver"
  port = 1433
  host = "sqlserver.local"
}

resource "cyral_repository_user_account" "t1" {
  repository_id = cyral_repository.t1.id
  name = "my-user-account"
  auth_scheme {
    aws_secrets_manager {
      secret_arn = "redacted"
    }
  }
}

resource "cyral_repository_access_rules" "t1" {
  repository_id = cyral_repository.t1.id
  user_account_id = cyral_repository_user_account.t1.id

  # First rule to be considered
  rule {
    identity {
      type = "username"
      name = "my-user-account"
    }
    config {
      policy_ids = []
    }
    valid_from = "2022-09-19T23:50:48.606Z"
    valid_until = "2022-10-10T00:00:00Z"
  }

  # Second rule to be considered
  rule {
    identity {
      type = "email"
      name = "fake-username@gmail.com"
    }
  }

  # ... and so on
  rule {
    identity {
      type = "group"
      name = "fake-group"
    }
    config {
      policy_ids = []
    }
  }
}
```

**Comment by ricardorey10** [comment]
@VictorGFM @Yowgf there is a mess going on with the branches in this PR. I have no idea why, but even though I was on top of Hannah's PR the base branch suddenly changed, and it's really messy.

Once Hannah merges her user accounts stuff I'll rebase on top of main and make sure that everything is fine.

**Pull request merged**
